[feat] add caching for CLI file transform

mellyeliu opened the pull request:
## What changed / motivation ?
This PR introduces a caching mechanism to the CLI file transform for performance.

1. Input and Output File Caching
   - Caches the input file hash, output file hash, and compiled CSS for each file
   - Ensures files are only recompiled if:
     - The input file content has changed
     - The output file content does not match the expected hash
2. Cache Storage
   - Stores cache data defaulting to `node_modules/.stylex-cache`, 
   - Will add fast follow for configurable directory (`cachePath`)
   - Each file's cache data includes:
     - `inputHash`: Hash of the input file content
     - `outputHash`: Hash of the compiled output file content
     - `collectedCSS`: The compiled CSS
3. Cache Integration
   - Before recompiling, the CLI checks the cache:
     - If hashes match, the cached CSS is reused
     - If not, the file is recompiled, and the cache is updated

## Testing

Added tests to ensure:
- Files are skipped when cache is valid
- Files are recompiled when inputs or outputs change
- Cache is correctly written and read from the specified path
## Pre-flight checklist

- [x] I have read the contributing guidelines
      [Contribution Guidelines](https://github.com/facebook/stylex/blob/main/.github/CONTRIBUTING.md)
- [x] Performed a self-review of my code

nmn commented on packages/cli/package.json:
```diff
@@ -12,7 +12,7 @@
     "build": "npm run build:cjs && npm run build:esm",
     "build-haste": "rewrite-imports -i src/ -o lib/",
-    "test": "jest"
+    "test": "NODE_ENV=test jest"
```
Please use the `cross-env` package to set the environment variable so it can also work on Windows. :(

nmn commented on packages/cli/__tests__/compile-stylex-folder-test.js:
```diff
@@ -196,3 +203,83 @@ describe('individual testing of util functions', () => {
+  });
+
+  test('first compilation populates the cache', async () => {
+    fs.mkdirSync(config.output, { recursive: true });
+    fs.mkdirSync(cachePath, { recursive: true });
```
NIT: Use the `async` versions of these functions so more tests can run in parallel.

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+ */
+import fs from 'fs';
+import path from 'path';
+import crypto from 'crypto';
```
TODO: check if `crypto` is faster than `murmur`hash which is what we use for hashing classNames etc.

We should use whatever is fastest because we don't care about the cryptographic properties of hashing algorithms that are hard to reverse-engineer.

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+import path from 'path';
+import crypto from 'crypto';
+// $FlowFixMe
+import { mkdirp } from 'mkdirp';
```
Maybe we replace `mkdirp` everywhere with:
```
fs.mkdirSync(config.output, { recursive: true });
```

?

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+
+// Default cache directory in `node_modules/.stylex-cache`
+export function getDefaultCachePath() {
+  return process.env.NODE_ENV === 'test'
+    ? path.join('node_modules', '.stylex-cache-test')
+    : path.join('node_modules', '.stylex-cache');
+}
```
Why do we need two different cache paths for test and prod?

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+}
+
+function getCacheFilePath(cachePath, filePath) {
+  const fileName = filePath.replace(/[\\/]/g, '__');
```
Can we change the filePath to be relative to the project root instead? (This can be a follow-up PR)

This would make the caches portable and re-usable on different machines.

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+}
+
+export function readCache(filePath) {
+  const cacheFile = getCacheFilePath(getDefaultCachePath(), filePath);
```
Do you have a way to override `getDefaultCachePath` already? If not, please add some comments that the "defaultCachePath` is currently the only cache path.

mellyeliu replied:
adding this in a fast follow!

nmn commented on packages/cli/src/cache.js:
```diff
@@ -0,0 +1,75 @@
+
+export function readCache(filePath) {
+  const cacheFile = getCacheFilePath(getDefaultCachePath(), filePath);
+  if (fs.existsSync(cacheFile)) {
+    return JSON.parse(fs.readFileSync(cacheFile, 'utf-8'));
+  }
+  return null;
```
Consider making all file read/write operations asynchronous if it's not too complicated. This will make the code more easy to change in the future to support additional ways to cache.

nmn commented on packages/cli/src/transform.js:
```diff
@@ -94,16 +100,56 @@ export async function compileFile(
+      oldOutputHash = null;
+    }
+  } else {
+    console.log(`[stylex] Output file does not exist: ${outputFilePath}`);
```
Don't log this here. We will already log when we generate this file.

nmn commented on packages/cli/src/transform.js:
```diff
@@ -94,16 +100,56 @@ export async function compileFile(
   const outputFilePath = path.join(config.output, filePath);
+  const cachePath = getDefaultCachePath();
+
+  const inputHash = computeHash(inputFilePath);
```
Yup. Making things async will be an improvement as you can compute the hash *and* read the cache *and* check if the output file exists, all in parallel.

nmn commented on packages/cli/src/transform.js:
```diff
@@ -94,16 +100,56 @@ export async function compileFile(
+    try {
+      oldOutputHash = computeHash(outputFilePath);
+    } catch (err) {
+      console.error(
```
Why do we need the `try-catch` here but not when the hash is being calculated for the input file?

mellyeliu replied:
whoops leftover from debugging, will rm

nmn commented on packages/cli/src/transform.js:
```diff
@@ -94,16 +100,56 @@ export async function compileFile(
+  if (
+    cacheData &&
+    cacheData.inputHash === inputHash &&
+    fs.existsSync(outputFilePath) &&
```
This computation is being repeated from above. (line 107)

nmn reviewed (approved):
This is a pretty solid implementation. Please cleanup the logging so it isn't so noisy at the very least.

There are improvements that can be made by making more of the code asynchronous, but nothing that should block this PR.

mellyeliu commented:
Addressed nits and async caching logic

To address in a follow up PR:
- Specifying a custom `cachePath`
- Testing on `murmurHash` vs `crypto`
- Make file path relative to project root
